packages/babel-register/src/loader.mjs

```diff
@@ -0,0 +1,287 @@
+    // same location. To determine how Node.js would interpret an arbitrary .js
+    // file, search up the file system for the nearest parent package.json file
+    // and read its "type" field.
+    const format = await getPackageType(url);
+    // When a hook returns a format of 'commonjs', `source` is be ignored.
+    // The .register call at the beginning of the file already takes care of
+    // the common-js file
+    if (format === "commonjs") {
+      return {
+        format,
+        shortCircuit: true,
+      };
+    }
+
+    const { source: rawSource, responseURL } = await nextLoad(url, {
+      ...context,
+      format,
+    });
+
+    const file = fileURLToPath(responseURL || url);
+    const result = compile(rawSource.toString(), file, true);
+
+    if (result !== null) {
+      return result;
+    }
+  }
+
+  // Let Node.js handle all other URLs.
+  return nextLoad(url);
+}
+
+export function globalPreload({ port }) {
+  port.onmessage = evt => {
+    if (evt.data.opts) {
+      global._BABEL_ESM_REGISTER.register(evt.data.client, evt.data.opts);
+    } else if (evt.data.revert) {
+      global._BABEL_ESM_REGISTER.revert();
+    }
+  };
+  port.unref();
+
+  const fnc = {
+    register(client, opts) {
+      port.postMessage({ client, opts });
+    },
+    revert() {
+      port.postMessage({ revert: true });
+    },
+  };
+  return `\
+    globalThis._BABEL_ESM_REGISTER = { ${fnc.register.toString()}, ${fnc.revert.toString()} };
+    port.unref();
+  `;
+}
+
+export function getGlobalPreloadCode() {
+  const fnc = {
+    register() {
+      throw new Error(
+        "Modyfing @babel/register on-demand is not supported in this version of node"
+      );
+    },
+    revert() {
+      throw new Error(
+        "Modyfing @babel/register on-demand is not supported in this version of node"
+      );
+    },
+  };
+  return `\
+    globalThis._BABEL_ESM_REGISTER = { ${fnc.register.toString()}, ${fnc.revert.toString()} };
+  `;
+}
+
+// backwards compat
+export async function getSource(url, context, next) {
+  return load(url, context, next);
+}
+
+export async function getFormat(url, context, next) {
+  if (flag[url]) {
+    const format = await getPackageType(url);
+    return {
+      format: format || "commonjs",
+    };
+  }
+  return next(url, context, next);
+}
+
+async function getPackageType(url) {
+  // `url` is only a file path during the first iteration when passed the
+  // resolved url from the load() hook
+  // an actual file path from load() will contain a file extension as it's
+  // required by the spec
+  // this simple truthy check for whether `url` contains a file extension will
+  // work for most projects but does not cover some edge-cases (such as
+  // extensionless files or a url ending in a trailing space)
+  const ext = extname(url);
+  if (ext === ".cjs") return "commonjs";
+  if (ext === ".mjs") return "module";
+  const isFilePath = !!ext;
```

Comment:
this is from the docs and is based on the simple assumption that directories don't have file extensions, i guess we should stat here (and maybe cache the stat call)